src/chain/shared_state.rs

```diff
@@ -596,22 +600,32 @@ impl SectionProofChain {
     }
 
     // Verify this proof chain against the given key infos.
-    pub fn check_trust<'a, 'b, I>(&'a self, their_key_infos: I) -> TrustStatus<'a>
+    pub fn check_trust<'a, I>(&self, their_key_infos: I) -> TrustStatus
     where
-        I: IntoIterator<Item = (&'b Prefix<XorName>, &'b SectionKeyInfo)>,
+        I: IntoIterator<Item = (&'a Prefix<XorName>, &'a SectionKeyInfo)>,
     {
         let last_prefix = self.last_public_key_info().prefix();
-        let known_key_infos: BTreeSet<_> = their_key_infos
-            .into_iter()
-            .filter(|&(pfx, _)| last_prefix.is_compatible(pfx))
-            .map(|(_, info)| info)
-            .collect();
+
+        let mut known_key_infos = BTreeSet::new();
+        let mut their_key_infos_count = 0;
+
+        for (prefix, info) in their_key_infos {
+            their_key_infos_count += 1;
+
+            if last_prefix.is_compatible(prefix) {
+                let _ = known_key_infos.insert(info);
+            }
+        }
+
+        if their_key_infos_count == 0 {
+            return TrustStatus::TheirKeyInfosEmpty;
```

Comment:
That does not seem right, we should know whether we can trust stuff or not and not call this function if we cannot.